Extension/src/LanguageServer/cppbuildTaskProvider.ts

```diff
@@ -0,0 +1,320 @@
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+import * as path from 'path';
+import * as vscode from 'vscode';
+import * as os from 'os';
+import * as util from '../common';
+import * as telemetry from '../telemetry';
+import { Client } from './client';
+import * as configs from './configurations';
+import * as ext from './extension';
+import * as fs from 'fs';
+import * as nls from 'vscode-nls';
+import * as cp from "child_process";
+
+const localize: nls.LocalizeFunc = nls.loadMessageBundle();
+export const failedToParseTasksJson: string = localize("failed.to.parse.tasks", "Failed to parse tasks.json, possibly due to comments or trailing commas.");
+
+export interface CppBuildTaskDefinition extends vscode.TaskDefinition {
+    type: string;
+    label: string;
+    command: string;
+    args: string[];
+    options: cp.ProcessEnvOptions | undefined;
+}
+
+export class CppBuildTaskProvider implements vscode.TaskProvider {
+    static CppBuildScriptType: string = 'cppbuild';
+    static CppBuildSourceStr: string = "C/C++";
+    private tasks: vscode.Task[] | undefined;
+
+    constructor() {}
+
+    public async provideTasks(): Promise<vscode.Task[]> {
+        if (this.tasks !== undefined) {
+            return this.tasks;
+        }
+        return this.getTasks(false, false);
+    }
+
+    // Resolves a task that has no [`execution`](#Task.execution) set.
+    public resolveTask(_task: vscode.Task): vscode.Task | undefined {
+        const execution: vscode.ProcessExecution | vscode.ShellExecution | vscode.CustomExecution | undefined = _task.execution;
+        if (execution === undefined) {
+            const definition: CppBuildTaskDefinition = <any>_task.definition;
+            return this.getTask(definition.command, false, false, definition.args ? definition.args : [], definition);
+        }
+        return undefined;
+    }
+
+    // Generate tasks to build the current file based on the user's detected compilers, the user's compilerPath setting, and the current file's extension.
+    public async getTasks(returnCompilerPath: boolean, appendSourceToName: boolean): Promise<vscode.Task[]> {
+        if (this.tasks !== undefined) {
+            return this.tasks;
+        }
+        this.tasks = [];
+        const editor: vscode.TextEditor | undefined = vscode.window.activeTextEditor;
+        if (!editor) {
+            return [];
+        }
+
+        const fileExt: string = path.extname(editor.document.fileName);
+        if (!fileExt) {
+            return [];
+        }
+
+        // Don't offer tasks for header files.
+        const fileExtLower: string = fileExt.toLowerCase();
+        const isHeader: boolean = !fileExt || [".hpp", ".hh", ".hxx", ".h", ".inl", ""].some(ext => fileExtLower === ext);
```

Comment:
I know you didn't write this part originally, but it looks like the native code (settings.h, line ~80-88) is also tracking ".h++", ".hp", ".i", ".ii", ".ino" and ".ipp" header extensions, and ".c++", ".cp", and ".tcc" source file extensions.  Could you add those here?   (It also associates ".idl" with C++, but idl isn't actually C++, so I'm not sure about that one.)

Reply:
I modify the header and source file set as below:
["*.hpp", "*.hh", "*.hxx", "*.h++", "*.hp", "*.h", "*.ii", "*.ino", "*.inl", "*.idl", ""]
["*.cpp", "*.cc", "*.cxx", "*.c++", "*.cp", "*.ipp", "*.tcc"]